websockets/test_client_server.py

```diff
@@ -442,18 +442,41 @@ def test_invalid_status_error_during_client_connect(self):
         self.assertEqual(exception.code, 403)
         self.stop_server()
 
-    @unittest.mock.patch('websockets.server.read_request')
-    def test_connection_error_during_opening_handshake(self, _read_request):
-        _read_request.side_effect = ConnectionError
+    def do_error_during_opening_handshake(self, exception_cls):
+        written = []
 
-        self.start_server()
+        class WriteCollectingProtocol(WebSocketServerProtocol):
+
+            def client_connected(self, reader, writer):
+                original_write = writer.write
+
+                def collecting_write(data):
+                    original_write(data)
+                    written.append(data)
+                writer.write = collecting_write
+                super().client_connected(reader, writer)
+
+            @asyncio.coroutine
+            def handshake(self, *args, **kwargs):
+                raise exception_cls()
+
+        self.start_server(klass=WriteCollectingProtocol)
         # Exception appears to be platform-dependent: InvalidHandshake on
         # macOS, ConnectionResetError on Linux. This doesn't matter; this
         # test primarily aims at covering a code path on the server side.
```

Comment:
Good suggestion. On inspection, it seems like we fully control those exception types on the client side (and so may be able to tighten the assertion), but I can't easily confirm on Linux as I use a Mac, too (and I don't know what the code did when the comment was written).

Really, what this test is checking is whether [this line](https://github.com/aaugustin/websockets/blob/ccc5d34f068e1e3e6b51c49aa0c69983d9cd104d/websockets/server.py#L89) is getting called on the server side:

    self.writer.write(response.encode())

 which is the only difference between the `ConnectionError` case and the non-`ConnectionError` case, aside from debug vs. warning logging levels. Can you confirm this is really how you want the code to behave?

Beyond that, like you say in the comment, the test is more for the server side, so checking the exception type on the client side seems more of a can't hurt / nice to be there. If we're interested in the client-side code path, it might deserve to be separated out into its own test if the code path isn't already covered. Or maybe there is a unique code path on the client for this case, and this is the easiest way to trigger it.